File: docs/js/state.js
/**
 * Состояние игры — единственный источник правды.
 *
 * Модель не помнит ни здоровья, ни инвентаря, ни времени: всё это живёт здесь
 * и подставляется в промпт каждый ход. Ответ модели содержит только
 * предложения изменений, которые проверяет rules.js.
 *
 * Состояние — обычный объект без методов, поэтому сохраняется одним
 * JSON.stringify и восстанавливается так же.
 */

// Версия схемы: по ней распознаются сейвы, записанные другой версией игры.
export const STATE_VERSION = 1;

export const MAX_HEALTH = 100;
export const MAX_FATIGUE = 100;
export const MINUTES_PER_DAY = 24 * 60;

export const TIMES_OF_DAY = ['утро', 'день', 'вечер', 'ночь'];

/** Пустое состояние со всеми полями на умолчаниях. */
export function emptyState() {
  return {
    version: STATE_VERSION,

    // Карточка мира — то, что сгенерировано на старте. По ходу игры не меняется.
    world_title: '',
    genre: '',
    backstory: '',
    world_rules: [],

    character_name: '',
    character_background: '',
    character_traits: [],
    character_goal: '',

    // Изменяемая часть.
    stats: { health: MAX_HEALTH, max_health: MAX_HEALTH, fatigue: 0 },
    location: { name: '', description: '' },
    inventory: [],
    elapsed_minutes: 0,
    time_of_day: 'утро',

    chronicle: [],
    history: [],
    suggested_actions: [],

    npcs: [],
    flags: {},

    game_over: false,
  };
}

/** Сколько ходов сделал игрок. Нулевая запись истории — первая сцена, она не ход. */
export function turnNumber(state) {
  return Math.max(0, state.history.length - 1);
}

export function day(state) {
  return Math.floor(state.elapsed_minutes / MINUTES_PER_DAY) + 1;
}

export function elapsedText(state) {
  const withinDay = state.elapsed_minutes % MINUTES_PER_DAY;
  const hours = Math.floor(withinDay / 60);
  const minutes = withinDay % 60;
  if (state.elapsed_minutes < 60) return `${minutes} мин`;
  if (!minutes) return `${hours} ч`;
  return `${hours} ч ${String(minutes).padStart(2, '0')} мин`;
}

export function inventoryText(state) {
  if (!state.inventory.length) return 'пусто';
  return state.inventory.map((item) => item.name).join(', ');
}

/** Превращает карточку мира со старта в игровое состояние. */
export function fromWorldCard(world) {
  const state = emptyState();
  const character = world.character || {};
  const location = world.location || {};
  const stats = world.stats || {};

  state.world_title = world.world_title || '';
  state.genre = world.genre || '';
  state.backstory = world.backstory || '';
  state.world_rules = [...(world.world_rules || [])];

  state.character_name = character.name || '';
  state.character_background = character.background || '';
  state.character_traits = [...(character.traits || [])];
  state.character_goal = character.goal || '';

  state.stats = {
    health: Number(stats.health ?? MAX_HEALTH),
    max_health: Number(stats.max_health ?? MAX_HEALTH),
    fatigue: Number(stats.fatigue ?? 0),
  };
  state.location = {
    name: location.name || '',
    description: location.description || '',
  };
  state.inventory = (world.inventory || []).map((item) => ({
    name: item.name || '',
    description: item.description || '',
  }));
  state.suggested_actions = [...(world.suggested_actions || [])];

  // Первая сцена — нулевой ход: игрок ещё ничего не делал.
  state.history.push({
    number: 0,
    player: '',
    narrative: world.opening_scene || '',
    local: false,
  });
  state.chronicle.push(
    `Начало: ${state.character_name} в месте «${state.location.name}». ` +
      `Цель: ${state.character_goal}`
  );
  return state;
}

/**
 * Восстановление из сейва.
 *
 * Незнакомые поля отбрасываются, недостающие остаются на умолчаниях: файл,
 * записанный другой версией игры, не должен ронять загрузку.
 */
export function fromSaved(data) {
  const state = emptyState();
  if (!data || typeof data !== 'object') return state;

  for (const key of Object.keys(state)) {
    if (!(key in data)) continue;
    const value = data[key];
    const fallback = state[key];
    if (Array.isArray(fallback)) {
      if (Array.isArray(value)) state[key] = value;
    } else if (fallback !== null && typeof fallback === 'object') {
      if (value && typeof value === 'object') state[key] = { ...fallback, ...value };
    } else if (typeof value === typeof fallback) {
      state[key] = value;
    }
  }

  // Поля, появившиеся позже нулевой версии, могли не сохраниться.
  state.history = state.history.map((turn) => ({
    number: turn.number ?? 0,
    player: turn.player ?? '',
    narrative: turn.narrative ?? '',
    local: Boolean(turn.local),
  }));
  return state;
}

File: game/core/state.py
"""Состояние игры — единственный источник правды.

Модель не помнит ни здоровья, ни инвентаря, ни времени: всё это живёт здесь и
подставляется в промпт каждый ход. Ответ модели содержит только предложения
изменений, которые проверяет game/core/rules.py.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass, field, fields
from typing import Any

# Версия схемы: понадобится на этапе 7, чтобы не ронять старые сейвы.
STATE_VERSION = 1

MAX_HEALTH = 100
MAX_FATIGUE = 100
MINUTES_PER_DAY = 24 * 60

TIMES_OF_DAY = ("утро", "день", "вечер", "ночь")


def _known_fields(cls, data: dict) -> dict:
    """Оставляет только те ключи, которые датакласс действительно знает."""
    names = {field.name for field in fields(cls)}
    return {key: value for key, value in (data or {}).items() if key in names}


def _build(cls, data):
    return cls(**_known_fields(cls, data or {}))


@dataclass
class Stats:
    health: int = 100
    max_health: int = MAX_HEALTH
    fatigue: int = 0


@dataclass
class Location:
    name: str = ""
    description: str = ""


@dataclass
class Item:
    name: str
    description: str = ""


@dataclass
class Turn:
    number: int
    player: str
    narrative: str
    # True — ответ сочинил код, а не модель (отказ по усталости или ранам).
    local: bool = False
    # Кто сделал ход: заполняется, когда за одной партией сидят несколько человек.
    player_name: str = ""


@dataclass
class NPC:
    """Персонаж, с которым игрок взаимодействовал.

    sympathy: -100 (готов навредить) .. +100 (предан). Число живёт здесь,
    модель получает его в промпте и предлагает только изменение.
    """

    name: str
    role: str = ""
    description: str = ""
    sympathy: int = 0
    first_seen_turn: int = 0
    last_seen_turn: int = 0
    scenes: int = 1
    important: bool = False
    # Находится ли в сцене прямо сейчас — от этого зависит показ в интерфейсе.
    present: bool = False


@dataclass
class GameState:
    version: int = STATE_VERSION

    # Карточка мира — то, что сгенерировал этап 2. Не меняется по ходу игры.
    world_title: str = ""
    genre: str = ""
    backstory: str = ""
    world_rules: list[str] = field(default_factory=list)

    character_name: str = ""
    character_background: str = ""
    character_traits: list[str] = field(default_factory=list)
    character_goal: str = ""

    # Изменяемая часть.
    stats: Stats = field(default_factory=Stats)
    location: Location = field(default_factory=Location)
    inventory: list[Item] = field(default_factory=list)
    elapsed_minutes: int = 0
    time_of_day: str = "утро"

    chronicle: list[str] = field(default_factory=list)
    history: list[Turn] = field(default_factory=list)
    suggested_actions: list[str] = field(default_factory=list)

    npcs: list[NPC] = field(default_factory=list)
    # Заполнятся на этапе 6 (бой).
    flags: dict[str, Any] = field(default_factory=dict)

    game_over: bool = False

    # ------------------------------------------------------------ свойства

    @property
    def turn_number(self) -> int:
        """Сколько ходов сделал игрок. Нулевая запись истории — первая сцена,
        она не ход."""
        return max(0, len(self.history) - 1)

    @property
    def day(self) -> int:
        return self.elapsed_minutes // MINUTES_PER_DAY + 1

    def elapsed_text(self) -> str:
        hours, minutes = divmod(self.elapsed_minutes % MINUTES_PER_DAY, 60)
        if self.elapsed_minutes < 60:
            return f"{minutes} мин"
        return f"{hours} ч {minutes:02d} мин" if minutes else f"{hours} ч"

    def inventory_text(self) -> str:
        if not self.inventory:
            return "пусто"
        return ", ".join(item.name for item in self.inventory)

    # ------------------------------------------------------- сериализация

    def to_dict(self) -> dict:
        data = asdict(self)
        data["turn_number"] = self.turn_number
        data["day"] = self.day
        data["elapsed_text"] = self.elapsed_text()
        return data

    @classmethod
    def from_dict(cls, data: dict) -> GameState:
        """Восстановление из сейва.

        Незнакомые поля отбрасываются, недостающие остаются на умолчаниях:
        сейв, записанный другой версией игры, не должен ронять загрузку.
        """
        data = dict(data)
        nested = ("stats", "location", "inventory", "history", "npcs")
        state = cls(
            **_known_fields(cls, {k: v for k, v in data.items() if k not in nested})
        )
        state.stats = _build(Stats, data.get("stats"))
        state.location = _build(Location, data.get("location"))
        state.inventory = [_build(Item, item) for item in data.get("inventory") or []]
        state.history = [_build(Turn, turn) for turn in data.get("history") or []]
        state.npcs = [_build(NPC, npc) for npc in data.get("npcs") or []]
        return state

    @classmethod
    def from_world_card(cls, world: dict) -> GameState:
        """Превращает карточку мира с этапа 2 в игровое состояние."""
        character = world.get("character", {})
        location = world.get("location", {})
        stats = world.get("stats", {})

        state = cls(
            world_title=world.get("world_title", ""),
            genre=world.get("genre", ""),
            backstory=world.get("backstory", ""),
            world_rules=list(world.get("world_rules", [])),
            character_name=character.get("name", ""),
            character_background=character.get("background", ""),
            character_traits=list(character.get("traits", [])),
            character_goal=character.get("goal", ""),
            stats=Stats(
                health=int(stats.get("health", MAX_HEALTH)),
                max_health=int(stats.get("max_health", MAX_HEALTH)),
                fatigue=int(stats.get("fatigue", 0)),
            ),
            location=Location(
                name=location.get("name", ""),
                description=location.get("description", ""),
            ),
            inventory=[
                Item(name=item.get("name", ""), description=item.get("description", ""))
                for item in world.get("inventory", [])
            ],
            suggested_actions=list(world.get("suggested_actions", [])),
        )
        # Первая сцена — нулевой ход: игрок ещё ничего не делал.
        state.history.append(
            Turn(number=0, player="", narrative=world.get("opening_scene", ""))
        )
        state.chronicle.append(
            f"Начало: {state.character_name} в месте «{state.location.name}». "
            f"Цель: {state.character_goal}"
        )
        return state
